fracmanager/storage_state.go

```
@@ -0,0 +1,142 @@
+	return m.save()
+}
+
+func (m *StateManager) init(defaultState StorageState) error {
+	err := m.load()
+	if os.IsNotExist(err) {
+		m.current = defaultState
+		return m.save()
+	}
+	return err
+}
+
+func (m *StateManager) load() error {
+	data, err := os.ReadFile(m.filePath)
+	if err != nil {
+		return err
+	}
+	var state StorageState
+	if err := json.Unmarshal(data, &state); err != nil {
+		return err
+	}
+
+	m.current = state
+	return nil
+}
+
+func (m *StateManager) save() error {
+	data, err := json.Marshal(m.current)
+	if err != nil {
+		return err
+	}
+	return atomicWrite(m.filePath, data, 0o644)
+}
+
+// atomicWrite safely writes data to file using atomic replacement pattern
+// Ensures data integrity with proper synchronization and cleanup
+func atomicWrite(path string, data []byte, perm os.FileMode) error {
+	// Create temporary file
+	tmpPath := path + ".tmp"
+	f, err := os.OpenFile(tmpPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, perm)
+	if err != nil {
+		return fmt.Errorf("create temp file: %w", err)
+	}
+
+	defer func() {
+		if f != nil {
+			f.Close() // Close file if still open
+		}
+		if err != nil {
+			os.Remove(tmpPath) // Remove temp file on error
+		}
+	}()
+
+	// Write data to temporary file
+	if _, err = f.Write(data); err != nil {
+		return fmt.Errorf("write data: %w", err)
+	}
+
+	// Sync data to disk (force write from cache)
+	if err = f.Sync(); err != nil {
+		return fmt.Errorf("sync data: %w", err)
+	}
+
+	// Close file handle
+	if err = f.Close(); err != nil {
+		return fmt.Errorf("close file: %w", err)
+	}
+	f = nil // Mark as closed so defer doesn't close again
+
+	// Atomic replacement using rename (atomic operation on most filesystems)
+	if err = os.Rename(tmpPath, path); err != nil {
+		return fmt.Errorf("rename file: %w", err)
+	}
+
+	// Sync parent directory
+	if err = syncDir(filepath.Dir(path)); err != nil {
+		return fmt.Errorf("sync dir: %w", err)
+	}
+
+	return nil
+}
+
+func syncDir(dir string) error {
```

There is `MustSyncPath` function in `util/fs.go`, which does the same (or if you want to return an error, `SyncPath` function also exists)